### ISSUE TITLE
Investigate JIT deoptimization sequence in context of BG compiler (or additional isolate) triggering deoptimization

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#45213]
Since efa1e184283379f5d2a770e2bfeee2bca64edcfd our background compiler is no longer bailing out on class finalization. Instead it will get mutators to a safepoint and then possibly deoptimize dependent code and mutator stack's frames executing such code. (/cc @aam @rmacnak-google )

We should look though whether this is actually working as intended.

One concern that comes to mind is the following: When BG compiler will stop mutators it uses the normal safepointing mechanism. That means the mutator will be at **any** safepoint in the runtime. If the top-of-stack function is going to be deoptimized - as a result of invalidating dependent code - we need to be actually deoptimize frames on the stack.

Right now some IR instructions can emit code that ends up calling the runtime - and possibly hit a safepoint there - without being marked as deoptimizable. Though if the deoptimization information is missing we have a problem.

For example:

We can go to runtime to throw an exception:
```c++
class ShiftUint32OpInstr : public ShiftIntegerOpInstr {
  virtual bool ComputeCanDeoptimize() const { return false; }
  virtual bool MayThrow() const { return true; }
}
```

We can go to runtime to allocate memory
```c++
class AllocateContextInstr : public TemplateAllocation<0, NoThrow> {
  virtual bool ComputeCanDeoptimize() const { return false; }
}
```
(side note: strange that `AllocateContextInstr` is marked as `NoThrow` - since we probably can throw `OutOfMemoryException`)

So it seems that we should consider any IR instruction that can end up in runtime as possibly needing deoptimization info.

/cc @mraleph @alexmarkov

**Comment by mraleph** [comment]
I think it is a fair assessment - now any instruction that does a non-leaf call needs to be a lazy-deopt aware.

**Comment by mkustermann** [comment]
As part of addressing this issue, I've added a new test has the following failures:
```
vm/dart_2/deopt/restart_call_on_deopt_regress_46070_test RuntimeError (expected Pass)
vm/dart_2/deopt/restart_call_on_deopt_regress_46070_test Crash (expected Pass)
```
on configurations
```
dartk-win-debug-x64
dartk-win-product-x64
dartk-win-release-ia32
dartk-win-release-x64
```

**Comment by mkustermann** [comment]
To sumarize the work that has landed to make sharing JITed code safe in context of deoptimization:

* [x] **Deopt pruning** We have added support for deopt environment pruning, to ensure moved/hoisted/... instructions that lazy-deopt will have the correct environment and also ensure consumed inputs have been removed from the lazy deopt-env
* [x] **Outline runtime calls** We moved all runtime calls out of optimized `EmitNative` into stubs using register ABI - thereby ensuring that the default lazy-deopt environment is correct (no pushed arguments to runtime call on the stack when returning via lazy-deopt)
* [x] **Added missing deopt info** We have added (deopt-id, env) to some instructions that were missing it (e.g. some allocation instructions, ...)
* [x] **Deopt Safepoint Operations**: We added support for non lazy-deopt runtime calls to ensure some runtime calls can safely cause GC without checking into deoptimization safepoints (this is used in boxing allocation slow paths to avoid causing deopt in places where we don't have deopt env)
* [x] **Deopt to before-call** We have added support for marking instructions as lazy-deopt to before-call (thereby re-trying the call). This is needed in inlining and call specialization where the replaced instructions could cause deopt.

So for now all known issues related to JIT + deoptimization have been fixed, therefore I'm going to close this issue.
